src/mod_register.erl

```diff
@@ -112,19 +129,26 @@ process_welcome_message(KVs) ->
     Subject = proplists:get_value(subject, KVs, ""),
     {Subject, Body}.
 
+%%%
+%%% Hooks and IQ handlers
+%%%
+
 -spec c2s_stream_features([exml:element()], mongooseim:host_type(), jid:lserver()) ->
           [exml:element()].
 c2s_stream_features(Acc, _HostType, _LServer) ->
     [#xmlel{name = <<"register">>,
             attrs = [{<<"xmlns">>, ?NS_FEATURE_IQREGISTER}]} | Acc].
 
-unauthenticated_iq_register(_Acc,
-                            Server, #iq{xmlns = ?NS_REGISTER} = IQ, IP) ->
+-spec unauthenticated_iq_register(empty, mongooseim:host_type(), jid:server(), jlib:iq(), 
+                                  {inet:ip_address(), inet:port_number()} | undefined) -> 
+    exml:element() | empty.
```

Comment:
not sure that `empty` can return, as I see here can be returned or `exml:element()` or `mongoose_acc:t().`. Please replace `empty` to `mongoose_acc:t()`.

Reply:
I took this return type from the spec of `c2s_unauthenticated_iq` function from `mongoose_hooks` module. That function passes `empty` as `Acc` when running the hook. It looks like this hook is used only in `ejabberd_c2s` where is a check if the result is `empty`.

Reply:
Maybe I should remove this spec?

Reply:
`unauthenticated_iq_register(empty...` for this case spec is correct I suppose. But I don't see where `empty` field is passed. Maybe then make sense add clause here eg 

```erlang
....;
unauthenticated_iq_register(empty = Acc, _HostType, _Server, _IQ, _IP) ->
    Acc.
```
?

Reply:
This function could get something else than `empty` as the first argument if there is a hook executed before it. Currently we don't have such hooks but this might change and it is possible to add more hooks in custom modules. For me the return values are OK, I would only extend the type of the first argument, see my suggestion.

Reply:
Then I confused. Why variable has name `Acc`? So, this can be `empty | exml_element() | mongoose_acc:t()` or we sure that this can be only `empty | exml_element()`? If this can be only `empty | exml_element()` then we need change the name of variable. Or maybe I miss something...

Reply:
It's an unfortunate name clash. This "Acc" is not related to `mongoose_acc` in any way and `ejabberd_c2s` clearly expects `empty | exml_element()` as the result.

Reply:
Got it, thanks.